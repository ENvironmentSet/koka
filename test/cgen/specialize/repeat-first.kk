// this gives a compiler error
// fun repeat-first(times : int, xs as Cons(x, _)) : exn list<_>
//   if times == 0 then
//     return Nil

//   Cons(x, repeat-first(unsafe-decreasing(times - 1), xs))

fun repeat-first(times : int, xs : list<_>) : list<_>
  if times == 0 then
    return Nil

  match xs
    Nil -> Nil
    Cons(x, _) -> Cons(x, repeat-first(unsafe-decreasing(times - 1), xs))

fun main()
  repeat-first(10, [1]).show.println