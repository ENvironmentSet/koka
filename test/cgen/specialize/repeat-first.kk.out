[1,1,1,1,1,1,1,1,1,1]
 
cgen/specialize/repeat-first/.lift000-main: (times : int) -> list<int>
cgen/specialize/repeat-first/main: () -> console ()
cgen/specialize/repeat-first/repeat-first: forall<a> (times : int, xs : list<a>) -> list<a>